### PR TITLE
asserts,image: support gadget tracks in the model assertion

### PR DESCRIPTION
**Pull request by mvo5**
The model assertion supports "kernel: snapName=trackName" already.

This PR adds support for "gadget: snapname=trackName" so that this
also works for gadget snaps and adds the coresponding policy checks
and tests in snapstate.

### CODE REVIEW

File: asserts/device_asserts.go [hunk 1]
```diff
@@ -73,25 +73,40 @@ func (mod *Model) Architecture() string {
 	return mod.HeaderString("architecture")
 }
 
+// snapWithTrack represents a snap that includes optional track
```
Comment (pedronis) [inline asserts/device_asserts.go]: I didn't notice that we still have code about kernel-track to remove here

File: image/image.go [hunk 2]
```diff
@@ -295,9 +295,9 @@ func MockTrusted(mockTrusted []asserts.Assertion) (restore func()) {
 	}
 }
 
-func makeKernelChannel(kernelTrack, defaultChannel string) (string, error) {
-	errPrefix := fmt.Sprintf("cannot use kernel-track %q from model assertion", kernelTrack)
-	kch, err := snap.ParseChannel(kernelTrack, "")
+func makeChannelFromTrack(track, defaultChannel string) (string, error) {
+	errPrefix := fmt.Sprintf("cannot use track %q from model assertion", track)
```
Comment (pedronis) [inline image/image.go]: seems we should say kernel or gadget as well in these errors

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -957,18 +957,23 @@ func canSwitchChannel(st *state.State, snapName, newChannel string) error {
...
-	nch, err := snap.ParseChannel(newChannel, "")
-	if err != nil {
-		return err
+	if snapName == model.Kernel() && model.KernelTrack() != "" {
+		nch, err := snap.ParseChannel(newChannel, "")
+		if err != nil {
+			return err
+		}
+		if nch.Track != model.KernelTrack() {
+			return fmt.Errorf("cannot switch from kernel-track %q to %q", model.KernelTrack(), nch.String())
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: we probably want s/kernel-track/kernel track/

File: overlord/snapstate/snapstate.go [hunk 4]
```diff
@@ -957,18 +957,23 @@ func canSwitchChannel(st *state.State, snapName, newChannel string) error {
...
 	}
-	if nch.Track != model.KernelTrack() {
-		return fmt.Errorf("cannot switch from kernel-track %q to %q", model.KernelTrack(), nch.String())
+	if snapName == model.Gadget() && model.GadgetTrack() != "" {
+		nch, err := snap.ParseChannel(newChannel, "")
+		if err != nil {
+			return err
+		}
+		if nch.Track != model.GadgetTrack() {
+			return fmt.Errorf("cannot switch from gadget-track %q to %q", model.GadgetTrack(), nch.String())
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: same for s/gadget-track/gadget track/

File: asserts/device_asserts.go (hunk 1 above)
Comment (mvo5) [reply asserts/device_asserts.go]: I can refactor this if you prefer that this goes away now, e.g. I could make return Gadget/Kernel return a SnapWithTrack (we may need a better name for this). WDYT?

**Comment by mvo5** [comment]
Splitting up for easier review

**Pull request closed without merging**